### ISSUE TITLE
[3.3.5] Current spellcast not being sent on object creation

### ISSUE THREAD

**Issue by Treeston** [TrinityCore/TrinityCore#17886]
3.3.5 f5cbc17

Repro:
1. Begin casting a long spell (ex.: crafting spells like [Frostweave Bag], which has a 25 second cast time in 3.3.5) on character A
2. Walk into vision range of character A using character B.

Expected behavior:
* Character B will see the spell character A is casting

Observed behavior:
* Character B will see character A in idle pose, with no current spell cast.

**Comment by Treeston** [comment]
It should count as object creation if the mob leaves vision range. You'll get a CreateObject for that.

**Comment by Shauren** [comment]
Object creation has no place for a spell cast. Check range on sending SMSG_SPELL_GO

**Comment by Treeston** [comment]
`SMSG_SPELL_GO` (assumedly) won't be sent to anything outside vision range. Should it?
If you enter vision range after the spell cast starts, what packets does the server send?

Sending it to anybody that could potentially enter your range on spell start makes no sense - potentially, you could have somebody hearthstoning in from a different continent. So it has to be sent on object creation, but how? Is there a `SMSG_SPELL_GO` right after the `SMSG_UPDATE_OBJECT`?

**Comment by r00ty-tc** [comment]
I see a couple of problems. First, the spell start (not go) isn't sent on visibility gained. I tested a hacky way to add that and it worked if you teleported or moved in really quickly. If you walked in, it still didn't work.

Reason being I think, as @Shauren  says. Range. If you stand at max range (just before you lose visibility) and target a player, you won't see their spell casts at all. That's why it still then doesn't show with the hackfix if you move in normally.

So, I added some proper code. To send it specifically to the player in the visibility update. It solved the above problem. But what then happens is that the cast bar will be wrong. If for example the player was casting a 10sec spell and you walk in 5secs in, it would show you as if they're casting a 5sec spell.

We need to implement SMSG_RESUME_CAST_BAR. When I added an implementation of this I would then get the cast bar with the correct time, but not the spell casting animation. Also if I then stopped the cast the observer wouldn't see the stopped cast.

So, I added start and then resume. This works correctly. However, it doesn't quite match up with sniffs. Sniffs seem to only send the resume. So, it feels like a hack but works.

I'm away this weekend. But if no-one else has any interest I'll look next week.

**Comment by r00ty-tc** [comment]
Something like https://github.com/r00ty-tc/TrinityCore/commit/33c192c05701d653a5657bf4eed7cafc04e3264a although, I've only done basic testing and as I say, the sending of spell start upon entry of player to visual range isn't how it used to be. Only the resume cast bar was sent.

Something for people to use as a base over the weekend if interested though.

**Comment by Treeston** [comment]
Hm, so what you're saying is that if we sent the same packets we see in sniffs, the behavior isn't the same as retail? :o

**Comment by r00ty-tc** [comment]
No, I'm suggesting that perhaps in 3.3.5 retail we never saw the spellcasting animation if re-entering range. Just the cast bar. But, no way to prove that unless someone can find a video showing the effect.

**Comment by Treeston** [comment]
Anecdotally, no way. Doubt I'd have proof of it anywhere though.

**Comment by r00ty-tc** [comment]
Unless, something in the aura or create packets set up the animation. Since what you always see when someone enters range of a casting unit (player or creature).

CreateObject block
Aura update block
Resume casting opcode.

So, I suppose it would have to be somewhere there.

**Comment by Treeston** [comment]
Isn't emote state sent in the create packet? Dunno, first thing that came to mind.

**Comment by Shauren** [comment]
Don't even think of adding any retarded hacks like that

**Comment by r00ty-tc** [comment]
HENCE WHY IT WAS A POST HERE AND NOT A FUCKING COMMIT TO CORE OR A FUCKING PR.

DO YOU EVEN KNOW WHAT THE FUCK A PROOF OF FUCKING CONCEPT IS?

**Comment by r00ty-tc** [comment]
https://gist.github.com/r00ty-tc/013d10cb71c7a52802ceac79b2813b9a

Contains the relevant elements of the sniff.

**Comment by Treeston** [comment]
@Shauren at what kind of distance though? I mean, if I hearthstone from Stormwind to Dalaran, surely I never got a `_GO` for a spell being cast in Dalaran?

**Comment by r00ty-tc** [comment]
Updated the gist (same link) to contain a more useful spell (firebolt from a pet)

**Comment by r00ty-tc** [comment]
Just another piece of the puzzle. UpdateFlags "Unknown1" is set for units specifically when there is a resume opcode sent after the create block. Now this introduces another uint32.. Here's where it gets weird.

When Unknown1 is set, for players, both the LowGUID and the extra uint32 value will be fixed like this:

>[4] Unk Int32: 3381653872 (0xC98FF170)
>[4] Low GUID: 3381653864 (0xC98FF168)

And that's true regardless of the player. e.g. the Low GUID no longer matches the Low GUID of the player. In other cases it does.

Now. I thought, what about creatures? Well a similar thing with a different value. Unknown1 flag set and:

>[11] Unk Int32: 3192910192 (0xBE4FF170)
>[11] Low GUID: 3192910184 (0xBE4FF168)

Again for ALL creatures regardless of their guid, it seems to be these values (I didn't check them all, just some)

So, first 2 bytes different, last 2 bytes the same.. 

Finally Pets. Unknown1 and

>[4] Unk Int32: 3229606256 (0xC07FE170)
>[4] Low GUID: 3229606248 (0xC07FE168)

Not worked out the why. I've just found the odd correlation. Maybe something about update blocks I don't know about.

**Comment by Treeston** [comment]
More values, pushing the lowguid further back?

**Comment by cocolino** [comment]
http://pastebin.com/erehAjrY where a1 is ClientResumeCastBar struct, this is for WOD. Hope it helps, this might shed some light on it for 7.x also.
